lib/si-cli/src/engine/podman_engine.rs

```diff
@@ -23,104 +67,610 @@ impl ContainerEngine for PodmanEngine {
     }
 
     async fn ping(&self) -> CliResult<()> {
-        let ping_info = self.podman.ping().await?;
-        dbg!(&ping_info);
+        self.podman.ping().await?;
         Ok(())
     }
 
     async fn missing_containers(&self) -> Result<Vec<String>, SiCliError> {
-        todo!()
+        let mut missing_containers = Vec::new();
+        let containers = self.downloaded_systeminit_containers_list().await?;
+
+        for name in CONTAINER_NAMES.iter() {
+            let required_container = format!("systeminit/{0}", name);
+            if !containers.iter().any(|c| {
+                c.repo_tags
+                    .iter()
+                    .all(|t| t.contains(&required_container.to_string()))
+            }) {
+                missing_containers.push(required_container.to_string());
+            }
+        }
+
+        Ok(missing_containers)
     }
 
-    async fn download_missing_containers(&self, _missing_containers: Vec<String>) -> CliResult<()> {
-        todo!()
+    async fn download_missing_containers(&self, missing_containers: Vec<String>) -> CliResult<()> {
+        let m = MultiProgress::new();
+        let sty = ProgressStyle::with_template(
+            "{spinner:.red} [{elapsed_precise}] [{wide_msg:.yellow/blue}]",
+        )
+        .unwrap()
+        .progress_chars("#>-");
+
+        let total_size = 100123123;
+
+        println!("Found {0} missing containers", missing_containers.len());
+
+        let mut spawned = Vec::new();
+        for missing_container in missing_containers {
+            let pb = m.add(ProgressBar::new(total_size));
+            pb.set_style(sty.clone());
+
+            let mut message = "Downloading ".to_owned();
+            message.push_str(missing_container.as_str());
+
+            let podman = self.podman.clone();
+
+            let h1 = tokio::spawn(async move {
+                let pull_opts = PullOpts::builder()
+                    // TODO: Can the docker.io/ prefix be omitted?
+                    .reference(format!("docker.io/{}:stable", missing_container))
+                    .build();
+                let images = podman.images();
+                let mut stream = images.pull(&pull_opts);
+                while let Some(pull_report) = stream.next().await {
+                    match pull_report {
+                        Ok(pull_report) => {
+                            if let Some(stream) = pull_report.stream {
+                                pb.set_message(stream.trim().to_owned());
```

Comment:
The `stream` field on [`LibpodImagesPullReport`](https://docs.rs/podman-api/latest/podman_api/models/struct.LibpodImagesPullReport.html) prints the current output of the pull operation.

For example:

```
Copying blob 7a0437f04f83 done  
Copying config 300e315adb done  
Writing manifest to image destination
```

This means, unlike Docker, there's no way to determine download progress. As such, I updated the progress bar to print messages instead of track download progress.

The end result looks like this:

```sh
Downloading systeminit/jaeger
Downloading systeminit/postgres
Downloading systeminit/nats
Downloading systeminit/otelcol
Downloading systeminit/council
Downloading systeminit/veritech
Downloading systeminit/pinga
Downloading systeminit/sdf
Downloading systeminit/web
⠁ [00:00:02] [Copying blob sha256:762d56e9a80536698ac3ea36211b605a77caf480a77ba7734fd0b9bfb3d33449                                                                                                        ]
⠁ [00:00:01] [Copying blob sha256:208df09768884d1ae8806dd90560977463ab9e731400cd839c1530d5cf205db6                                                                                                        ]
⠁ [00:00:00] [Copying blob sha256:69b469d8248cd4b52f3f037c7b21bd48d54e50530823db48b861c3eb24c085b9                                                                                                        ]
⠁ [00:00:00] [Copying blob sha256:3ac90fd015c1e294c1b4d9eb71bc9d04330221e73ca7270f8bc9a3571fcfbb09                                                                                                        ]
⠁ [00:00:00] [Copying blob sha256:8742cdf797e70982869018e441c402c440a1315e98fc3fef45caca328ada3929                                                                                                        ]
All containers successfully downloaded
Starting systeminit/jaeger:stable as local-jaeger-1
Starting systeminit/postgres:stable as local-postgres-1
Starting systeminit/nats:stable as local-nats-1
Starting systeminit/otelcol:stable as local-otelcol-1
Starting systeminit/council:stable as local-council-1
Starting systeminit/veritech:stable as local-veritech-1
Starting systeminit/pinga:stable as local-pinga-1
Starting systeminit/sdf:stable as local-sdf-1
Starting systeminit/web:stable as local-web-1
All system components running... System Initiative is alive!
```